File: examples/calculator-java-hexagonal-cqrs/src/main/java/com/oopforge/example/calculator/application/query/HistoryQueryRepository.java
package com.oopforge.example.calculator.application.query;

import com.oopforge.example.calculator.domain.CalculationId;
import java.util.List;
import java.util.Optional;

/**
 * Outbound read port (query side) — returns read models, never aggregates.
 */
public interface HistoryQueryRepository {

    List<HistorySummary> listRecent(int limit);

    Optional<HistorySummary> findById(CalculationId calculationId);
}

File: examples/calculator-python-hexagonal-cqrs/app/application/query/history_query_repository.py
from typing import Protocol

from app.application.query.history_summary import HistorySummary
from app.domain.calculation.value import CalculationId


class HistoryQueryRepository(Protocol):
    """Read-side (query) port — returns read models, never aggregates."""

    def list_recent(self, limit: int) -> list[HistorySummary]: ...

    def find_by_id(self, calculation_id: CalculationId) -> HistorySummary | None: ...
